x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java

```diff
@@ -324,6 +327,35 @@ private LogicalPlan parse(String query, QueryParams params) {
         return parsed;
     }
 
+    static void handleFieldCapsFailures(
+        boolean allowPartialResults,
+        EsqlExecutionInfo executionInfo,
+        Map<String, List<FieldCapabilitiesFailure>> failures
+    ) throws Exception {
+        FailureCollector failureCollector = new FailureCollector();
+        for (var e : failures.entrySet()) {
+            String clusterAlias = e.getKey();
+            EsqlExecutionInfo.Cluster cluster = executionInfo.getCluster(clusterAlias);
+            if (cluster.getStatus() != EsqlExecutionInfo.Cluster.Status.RUNNING) {
+                continue;
+            }
+            if (allowPartialResults == false && executionInfo.isSkipUnavailable(clusterAlias) == false) {
+                for (FieldCapabilitiesFailure failure : e.getValue()) {
+                    failureCollector.unwrapAndCollect(failure.getException());
+                }
+            } else if (cluster.getFailures().isEmpty()) {
+                executionInfo.swapCluster(clusterAlias, (k, curr) -> {
+                    var shardFailures = e.getValue().stream().map(f -> new ShardSearchFailure(f.getException())).toList();
+                    return new EsqlExecutionInfo.Cluster.Builder(cluster).addFailures(shardFailures).build();
```

Comment:
Is wrapping into ShardSearchFailure necessary here? If so, it would be nice to add source index/shard to it

Reply:
Yes, that's existing behavior. ++ to add the shard target to the failure - see https://github.com/elastic/elasticsearch/pull/130840/commits/406f8b225f5ad15ea3ec23c90075e9656c854d1b.